solr/core/src/java/org/apache/solr/filestore/FileStore.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.filestore;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.function.BiConsumer;
+import java.util.function.Predicate;
+
+import org.apache.solr.common.MapWriter;
+import org.apache.solr.filestore.FileStoreAPI.MetaData;
+
+
+public interface FileStore {
+
+  /**Store a file into the filestore. This should ensure that it is replicated
+   * across all nodes in the cluster
+   */
+  void put(String path , MetaData metadata, ByteBuffer filecontent) throws IOException;
```

Comment:
This contract for putting a file assumes the entire file can fit in memory. However, for very large files, it might be more preferable to stream the files in. Even for a package store, this just means larger memory footprint. Can we use an OutputStream instead?

Reply:
yeah, it is doable. But for relatively smaller files, It's better to keep it in memory because it can be served from memory . This is a tradeoff

Reply:
In that case, the interface could account for both usecases (in memory and streamed) if it just used OutputStream instead of ByteBuffer.